File: apps/ios/OYBC/Services/RecurringBoardTemplates.swift
import Foundation

// MARK: - Recurring board templates — Phase 6.2 (Preset-pool)
//
// Pure functions over `RecurringBoardTemplate[]` + `Board[]` + `Task[]`
// that mirror `packages/shared/src/algorithms/recurringBoardTemplates.ts`.
// No persistence, no platform-specific code beyond Foundation. Both
// platforms must keep these in lockstep — when the TS version changes,
// mirror it here in the same PR.
//
// Canonical design: docs/ARCHITECTURE.md §Phase 6.2.

/// One pending template + the window it's pending for. Consumed by the
/// platform spawn driver.
struct PendingTemplateSpawn {
    let template: RecurringBoardTemplate
    /// Local ISO8601 from `wizardLocalISOString(window.start)`.
    let windowStart: String
    /// Local ISO8601 from `wizardLocalISOString(window.end)`.
    let windowEnd: String
    /// Default name for the spawned board: "<template name> — <window label>".
    let suggestedName: String
}

/// Outcome of `validateSpawnPool`. Mirrors the TS-side `SpawnPoolValidation`.
enum SpawnPoolValidation: Equatable {
    case ok
    case failure(SpawnPoolFailureReason)
}

/// Pure-validation failure reasons returned by `validateSpawnPool`.
/// Strict mirror of TS `SpawnPoolFailureReason` in
/// `packages/shared/src/algorithms/recurringBoardTemplates.ts`.
///
/// Driver-only outcomes (`noPoolTasksResolved`, `spawnFailed`) live on
/// `SpawnAttentionReason` instead — keeping this enum a strict mirror
/// of the shared contract means a future cross-platform schema check
/// can compare the two enums as sets without filtering out platform
/// drift.
enum SpawnPoolFailureReason: String, Equatable {
    case poolTooSmall = "pool_too_small"
    case hasDeletedTasks = "has_deleted_tasks"
    case unsupportedTimeframe = "unsupported_timeframe"
    case unsupportedCenter = "unsupported_center"
}

/// Attention reason surfaced on a template row when the spawn driver
/// skipped or failed for a particular template. Superset of
/// `SpawnPoolFailureReason` plus the two driver-only outcomes:
/// `noPoolTasksResolved` (the resolved pool was empty — every seed
/// task got soft-deleted) and `spawnFailed` (the txn threw
/// unexpectedly).
///
/// Mirrors the TS-side union
/// `SpawnPoolFailureReason | 'no_pool_tasks_resolved' | 'spawn_failed'`
/// used in `useRecurringBoardSpawn.ts`'s `RecurringSpawnDigest`.
enum SpawnAttentionReason: String, Equatable {
    case poolTooSmall = "pool_too_small"
    case hasDeletedTasks = "has_deleted_tasks"
    case unsupportedTimeframe = "unsupported_timeframe"
    case unsupportedCenter = "unsupported_center"
    case noPoolTasksResolved = "no_pool_tasks_resolved"
    case spawnFailed = "spawn_failed"
    /// Board Sources P3 — a pulled board-kind source's board is deleted
    /// or archived. Unlike an EMPTY source (contributes nothing, never
    /// blocks), this blocks the window and the Boards tab ASKS the user
    /// (docs/BOARD_SOURCES.md §Boards as sources).
    case sourceBoardMissing = "source_board_missing"
}

extension SpawnAttentionReason {
    /// Lifts a pure validation failure into the wider attention enum.
    /// The wider enum is a strict superset, so the rawValue lookup is
    /// non-failable in practice; the `!` is documenting that invariant.
    init(_ failure: SpawnPoolFailureReason) {
        self = SpawnAttentionReason(rawValue: failure.rawValue)!
    }
}

/// Counts the cells that need a Task placement for a given board configuration.
/// Even-sized boards have no center; odd-sized boards with a FREE
/// center reserve one cell for the auto-completed free space; otherwise the
/// center cell gets a regular task.
func recurringTemplateFillableCellCount(
    boardSize: Int,
    centerSquareType: CenterSquareType
) -> Int {
    let total = boardSize * boardSize
    let hasCenter = boardSize % 2 == 1
    let centerOmitsTask = hasCenter && centerSquareType == .free
    return total - (centerOmitsTask ? 1 : 0)
}

/// Detects templates pending a spawn for the current window.
///
/// A template is pending when ALL of the following hold:
///   - `!isDeleted` AND `isActive`.
///   - `timeframe` is one of daily/weekly/monthly/yearly (custom excluded).
///   - The current window's `startDate` differs from `lastSpawnedWindowKey`.
///   - **Idempotency belt**: there is no existing non-deleted Board with
///     matching `spawnedFromTemplateId + startDate`.
///
/// Output ordering is the SPAWN order the caller must iterate: parents first
/// (yearly → monthly → weekly → daily, stable within a tier), then
/// dependency-ordered so a template spawns after every pending template whose
/// series it pulls from (cycles fall back to parents-first).
///
/// - Parameters:
///   - templates: All non-deleted templates for the active user.
///   - boards: All boards for the active user (the idempotency belt, and
///     resolving board sources to their series for the dependency order).
///   - weekStartDay: Controls weekly window boundaries (`"monday"` / `"sunday"`).
///   - now: Reference date for window computation.
func findTemplatesPendingSpawn(
    templates: [RecurringBoardTemplate],
    boards: [Board],
    weekStartDay: String,
    now: Date
) -> [PendingTemplateSpawn] {
    var pending: [PendingTemplateSpawn] = []

    for template in templates {
        if template.isDeleted { continue }
        if !template.isActive { continue }
        if template.timeframe == .custom { continue }

        guard let window = computeTimeframeBoundaries(
            timeframe: template.timeframe,
            referenceDate: now,
            weekStartDay: weekStartDay
        ) else { continue } // .custom — already filtered above, defensive

        let startISO = wizardLocalISOString(window.start)
        let endISO = wizardLocalISOString(window.end)

        if template.lastSpawnedWindowKey == startISO { continue }

        // Idempotency belt — see header doc.
        let alreadySpawned = boards.contains { board in
            !board.isDeleted &&
            board.spawnedFromTemplateId == template.id &&
            board.startDate == startISO
        }
        if alreadySpawned { continue }

        pending.append(PendingTemplateSpawn(
            template: template,
            windowStart: startISO,
            windowEnd: endISO,
            suggestedName: deriveSpawnedBoardName(template: template, windowStart: startISO)
        ))
    }

    // Series binding (loose-ends sweep 2026-09-09) — PARENTS SPAWN FIRST
    // (yearly → monthly → weekly → daily; stable within a tier) as the BASE
    // order, then DEPENDENCY-ORDERED (final-review F3): a template spawns
    // after every pending template whose series it pulls from, so its board
    // source binds to the series' instance open NOW — the one this pass just
    // spawned — instead of resolving `.noWindow`. TS twin:
    // `findTemplatesPendingSpawn`'s tail sort + `orderBySourceDependency`.
    func tier(_ timeframe: Timeframe) -> Int {
        switch timeframe {
        case .yearly: return 0
        case .monthly: return 1
        case .weekly: return 2
        case .daily: return 3
        default: return 4
        }
    }
    let parentsFirst = pending.enumerated().sorted { a, b in
        let ta = tier(a.element.template.timeframe)
        let tb = tier(b.element.template.timeframe)
        if ta != tb { return ta < tb }
        return a.offset < b.offset
    }.map { $0.element }
    return orderBySourceDependency(parentsFirst, boards: boards)
}

/// Stable topological order over board-source → series edges between the
/// pending spawns (final-review F3). An edge `S → T` exists when pending
/// template `T` has a `.board` source whose stored board belongs to pending
/// template `S`'s series (`spawnedFromTemplateId == S.id`). A self-edge (a
/// template pulling its own series) is ignored; a one-off source (no
/// `spawnedFromTemplateId`) or a series not pending in this pass adds no edge.
///
/// Kahn-style and stable: each step emits the FIRST entry in `base` order
/// whose dependencies are all emitted. When none is ready (a cycle), the
/// first remaining entry in `base` order is emitted — the cycle falls back to
/// the parents-first order, never traps. Mirrors the TS
/// `orderBySourceDependency`.
///
/// - Parameters:
///   - base: Pending spawns in parents-first order.
///   - boards: The user's boards (resolve a stored source id to its series).
/// - Returns: The same entries, dependency-ordered.
private func orderBySourceDependency(
    _ base: [PendingTemplateSpawn],
    boards: [Board]
) -> [PendingTemplateSpawn] {
    let pendingIds = Set(base.map { $0.template.id })
    var seriesByBoardId: [String: String] = [:]
    for b in boards {
        if let series = b.spawnedFromTemplateId, !series.isEmpty { seriesByBoardId[b.id] = series }
    }
    var deps: [String: Set<String>] = [:]
    for p in base {
        var own = Set<String>()
        for source in p.template.sources ?? [] where source.kind == .board {
            if let series = seriesByBoardId[source.sourceId],
               series != p.template.id, pendingIds.contains(series) {
                own.insert(series)
            }
        }
        deps[p.template.id] = own
    }
    var remaining = base
    var emitted = Set<String>()
    var ordered: [PendingTemplateSpawn] = []
    while !remaining.isEmpty {
        let pick = remaining.firstIndex { p in
            (deps[p.template.id] ?? []).allSatisfy { emitted.contains($0) }
        } ?? 0 // cycle → parents-first fallback
        let next = remaining.remove(at: pick)
        emitted.insert(next.template.id)
        ordered.append(next)
    }
    return ordered
}

/// "<template name> — <window label>" helper.
///
/// Takes the local ISO8601 string for parity with the TS counterpart
/// (`packages/shared/src/algorithms/recurringBoardTemplates.ts`'s
/// `deriveSpawnedBoardName(template, windowStart: string)`). Both
/// platforms now consume the same wire shape — a future change to
/// either side's date format won't silently desync the other.
func deriveSpawnedBoardName(template: RecurringBoardTemplate, windowStart: String) -> String {
    let trimmed = template.name.trimmingCharacters(in: .whitespacesAndNewlines)
    guard let startDate = parseISO8601Date(windowStart) else {
        return trimmed.isEmpty ? "" : trimmed
    }
    // Absolute, not relative: this string is PERSISTED as the board's name.
    let label = formatWindowLabel(timeframe: template.timeframe, startDate: startDate)
    if trimmed.isEmpty { return label }
    return "\(trimmed) — \(label)"
}

/// Validates a template's pool. Mirrors the TS-side `validateSpawnPool`.
///
/// Pool semantics: `poolTasks.count >= required`. The spawn shuffles +
/// slices, so any extras become the random subset. The earlier
/// strict-fit `.all` strategy was dropped during the Phase 6.2 UX
/// rework — it was a special case of loose-fit where the user picked
/// exactly `required`.
func validateSpawnPool(
    template: RecurringBoardTemplate,
    poolTasks: [Task]
) -> SpawnPoolValidation {
    if template.timeframe == .custom {
        return .failure(.unsupportedTimeframe)
    }
    if template.centerSquareType == .chosen {
        return .failure(.unsupportedCenter)
    }
    if poolTasks.contains(where: { $0.isDeleted }) {
        return .failure(.hasDeletedTasks)
    }

    // Reject duplicate task ids in the resolved pool. Mirrors the
    // TS-side validateSpawnPool dedupe check. See its rationale.
    let uniqueIds = Set(poolTasks.map { $0.id })
    if uniqueIds.count != poolTasks.count {
        return .failure(.poolTooSmall)
    }

    let required = recurringTemplateFillableCellCount(
        boardSize: template.boardSize,
        centerSquareType: template.centerSquareType
    )
    if poolTasks.count < required { return .failure(.poolTooSmall) }
    return .ok
}

/// Builds the cell-by-cell placement for a spawn. Returns an array of
/// length `boardSize²` where each entry is a Task to place on that cell,
/// or `nil` for the auto-completed FREE center on odd-sized
/// boards.
///
/// Thin wrapper over the shared `BoardPlacement.placeBoard` core — mirrors
/// the TS-side `buildSpawnPlacement` (which now delegates to bingo-core's
/// `placeBoard`). Resolves the template's geometry / randomization / pool,
/// then delegates the cell walk. The spawn path never pins a CHOSEN center
/// (`validateSpawnPool` rejects CHOSEN templates), so no `chosenCenterId`
/// is passed — a CHOSEN center that slipped through is treated as ordinary,
/// exactly as the old hand-rolled loop did.
///
/// Optional `rng` parameter for deterministic test placement; passing a
/// seeded closure produces a stable order.
///
/// `validateSpawnPool` MUST be called first — this function's behavior is
/// undefined for a pool that fails validation.
func buildSpawnPlacement(
    template: RecurringBoardTemplate,
    poolTasks: [Task],
    rng: @escaping () -> Double = { Double.random(in: 0..<1) }
) -> [Task?] {
    BoardPlacement.placeBoard(
        items: poolTasks,
        gridSize: template.boardSize,
        centerType: template.centerSquareType,
        randomize: template.isRandomized,
        rng: rng
    )
}

// MARK: - "Repeat this board…" (Task Pools + Recurring Boards Rework, P6)
//
// docs/POOLS_RECURRING.md §Surfaces item 7 — a one-off board's Board screen
// offers `↻ Repeat this board…`, which mints a NEW spawn record from the
// board's current live tasks (as `manualTaskIds`, an own-mix repeating
// board with zero pools) and back-stamps the board's
// `spawnedFromTemplateId`. Unlike Phase 6.2's fresh-create spawn path, this
// does NOT spawn a new board immediately — the board being repeated IS this
// window's board, so an immediate spawn would double it up.

/// The fields needed to construct a new `RecurringBoardTemplate` from an
/// existing one-off board. Pure value object — `AppDatabase.repeatBoardAsTemplate`
/// is responsible for minting the id/timestamps and performing the write.
struct RepeatBoardTemplateInput {
    let name: String
    /// The CHOSEN cadence, NOT `board.timeframe` — a repeating board's
    /// cadence becomes its timeframe going forward.
    let timeframe: Timeframe
    let boardSize: Int
    let centerSquareType: CenterSquareType
    let isRandomized: Bool
    let manualTaskIds: [String]
    /// Always `true` — a freshly-repeated board starts active.
    let isActive: Bool
    let lastSpawnedWindowKey: String
    /// Decode-compat snapshot, mirrors `manualTaskIds` verbatim (see
    /// `RecurringBoardTemplate`'s `seedTaskIds` doc — never read live).
    let seedTaskIds: [String]
    /// Always empty — an own-mix repeating board pulls no pools.
    let poolIds: [String]
    /// Always empty — nothing to remove from an empty pool union.
    let removedTaskIds: [String]
}

/// Builds the `RepeatBoardTemplateInput` for "Repeat this board…".
///
/// `lastSpawnedWindowKey` is computed against the CADENCE's window
/// (`computeTimeframeBoundaries(timeframe: cadence, ...)`), never the
/// board's own `timeframe` — a one-off Tuesday `.daily` board repeated
/// `.weekly` must key off that Tuesday's WEEK window, not the day. Getting
/// this wrong silently double-spawns mid-window on the next Boards-tab
/// open (`findTemplatesPendingSpawn` would see a stale key and fire).
///
/// - Parameters:
///   - boardName: The source board's name — carried forward as the
///     template's name.
///   - boardSize: The source board's grid size.
///   - centerSquareType: The source board's center type.
///   - isRandomized: The source board's shuffle setting.
///   - boardStartDate: The source board's `startDate` (ISO8601) — the
///     anchor for locating the cadence window that contains it.
///   - boardTaskIds: Caller-resolved distinct, non-deleted, non-FREE-center
///     task ids currently placed on the board, in placement order.
///   - cadence: The user-chosen repeat cadence (Daily/Weekly/Monthly/Yearly
///     — `.custom`/`.indefinite` never reach here; the picker doesn't offer
///     them).
///   - weekStartDay: `"monday"` / `"sunday"` — only affects `.weekly`.
/// - Returns: `nil` only on an unparseable `boardStartDate` (should never
///   happen for a live board).
func buildRepeatBoardTemplateInput(
    boardName: String,
    boardSize: Int,
    centerSquareType: CenterSquareType,
    isRandomized: Bool,
    boardStartDate: String,
    boardTaskIds: [String],
    cadence: Timeframe,
    weekStartDay: String
) -> RepeatBoardTemplateInput? {
    guard let anchor = parseISO8601Date(boardStartDate) else { return nil }
    guard let window = computeTimeframeBoundaries(
        timeframe: cadence,
        referenceDate: anchor,
        weekStartDay: weekStartDay
    ) else { return nil } // cadence is never .custom/.indefinite in practice — defensive

    let windowKey = wizardLocalISOString(window.start)

    return RepeatBoardTemplateInput(
        name: boardName,
        timeframe: cadence,
        boardSize: boardSize,
        centerSquareType: centerSquareType,
        isRandomized: isRandomized,
        manualTaskIds: boardTaskIds,
        isActive: true,
        lastSpawnedWindowKey: windowKey,
        seedTaskIds: boardTaskIds,
        poolIds: [],
        removedTaskIds: []
    )
}

File: packages/shared/src/algorithms/recurringBoardTemplates.ts
/**
 * Recurring board templates — Phase 6.2 (Preset-pool Recurring Boards)
 *
 * Pure functions over `RecurringBoardTemplate[]` + `Board[]` + `Task[]` that:
 *
 *   1. Detect which templates are pending a spawn for the current window
 *      (`findTemplatesPendingSpawn`). The Boards-tab `onAppear` hook calls
 *      this alongside Phase 6.1's `findPendingRecurringBoards`; templates
 *      spawn directly to the board list (no banner — locked design).
 *
 *   2. Validate that a template's pool is well-formed (`validateSpawnPool`).
 *      Returns a structured failure reason so the platform spawn driver
 *      can surface a "needs attention" indicator on the template row
 *      instead of silently failing.
 *
 *   3. Build the cell-by-cell placement of `Task | null` for a spawn
 *      (`buildSpawnPlacement`). Mirrors the wizard's `buildWizardPlacement`
 *      but driven by template config + pool, with optional RNG injection
 *      for deterministic tests.
 *
 * No persistence; no platform code; no side effects. The platform spawn
 * driver wraps the multi-table write (Board + BoardTasks + template
 * `lastSpawnedWindowKey` update) in a single transaction.
 *
 * Canonical design: docs/ARCHITECTURE.md §Phase 6.2.
 */

import { placeBoard, fillableCellCount, CenterSquareType, type BoardSize } from '@oybc/bingo-core';
import { getTimeframeBoundaries, formatTimeframeLabel } from './calendarBoundaries';
import { Timeframe } from '../constants/enums';
import type { Board } from '../types/board';
import type { Task } from '../types/task';
import type { RecurringBoardTemplate } from '../types/recurringBoardTemplate';
import type { WeekStartDay } from './calendarBoundaries';

/**
 * One pending template + the window it's pending for. The platform spawn
 * driver consumes this to drive a Board+BoardTasks insert plus a template
 * `lastSpawnedWindowKey` update inside a single transaction.
 */
export interface PendingTemplateSpawn {
  template: RecurringBoardTemplate;
  /** Local ISO8601 from `getTimeframeBoundaries()`. */
  windowStart: string;
  /** Local ISO8601 from `getTimeframeBoundaries()`. */
  windowEnd: string;
  /**
   * Default name to put on the spawned board (template.name + window label,
   * e.g. "Daily Workout — May 7, 2026"). The spawn driver may override.
   */
  suggestedName: string;
}

/**
 * Outcome of `validateSpawnPool`. The platform spawn driver branches on
 * `ok` — if `ok=false`, skip the spawn for this window AND surface the
 * `reason` via a "needs attention" indicator on the template's row in
 * the Profile recurring-templates list.
 */
export type SpawnPoolValidation =
  | { ok: true }
  | { ok: false; reason: SpawnPoolFailureReason };

export type SpawnPoolFailureReason =
  | 'pool_too_small'         // pool < fillable cells
  | 'has_deleted_tasks'      // any task in the resolved pool has isDeleted=true
  | 'unsupported_timeframe'  // template.timeframe === CUSTOM (form should prevent this)
  | 'unsupported_center';    // centerSquareType === CHOSEN (MVP excludes this)

// `fillableCellCount` now lives in @oybc/bingo-core (imported above) — the
// same primitive drives the wizard, the spawn path, and Play's server-side
// board fan-out. Kept as a plain re-import here (not re-exported) so existing
// `recurringBoardTemplates` callers see no change.

/**
 * Detects templates pending a spawn for the current window.
 *
 * A template is pending when ALL of the following hold:
 *   - `!isDeleted` AND `isActive`.
 *   - `timeframe` is one of DAILY/WEEKLY/MONTHLY/YEARLY (CUSTOM excluded).
 *   - The current window's `startDate` (computed via `getTimeframeBoundaries`)
 *     differs from `lastSpawnedWindowKey` — i.e. we haven't spawned this
 *     window yet according to template state.
 *   - **Idempotency belt**: there is no existing non-deleted Board with
 *     `spawnedFromTemplateId === template.id` AND `startDate === windowStart`.
 *     This is the secondary check; it catches the rare case where
 *     `lastSpawnedWindowKey` lags behind reality (e.g., the previous spawn
 *     succeeded but the template-update step in the same txn was rolled
 *     back). With both checks, single-device idempotency is structurally
 *     guaranteed; the multi-device race remains an accepted MVP limitation.
 *
 * Output ordering is the input order — the caller may sort however it wants
 * (typically by `template.updatedAt desc` for UI display).
 *
 * @param templates - All non-deleted templates for the active user.
 * @param boards - All boards for the active user (used for the idempotency belt).
 * @param weekStartDay - Controls weekly window boundaries.
 * @param now - Reference date for window computation (typically `new Date()`).
 */
export function findTemplatesPendingSpawn(
  templates: RecurringBoardTemplate[],
  boards: Board[],
  weekStartDay: WeekStartDay,
  now: Date,
): PendingTemplateSpawn[] {
  const pending: PendingTemplateSpawn[] = [];

  for (const template of templates) {
    if (template.isDeleted) continue;
    if (!template.isActive) continue;
    if (template.timeframe === Timeframe.CUSTOM) continue;

    const { startDate, endDate } = getTimeframeBoundaries(
      template.timeframe,
      now,
      weekStartDay,
    );

    if (template.lastSpawnedWindowKey === startDate) continue;

    // Idempotency belt — see doc above.
    const alreadySpawned = boards.some(
      (b) =>
        !b.isDeleted &&
        b.spawnedFromTemplateId === template.id &&
        b.startDate === startDate,
    );
    if (alreadySpawned) continue;

    pending.push({
      template,
      windowStart: startDate,
      windowEnd: endDate,
      suggestedName: deriveSpawnedBoardName(template, startDate),
    });
  }

  return pending;
}

/**
 * Composes a default name for a spawned board: "<template name> — <window label>".
 * Used by `findTemplatesPendingSpawn`; the spawn driver may override.
 */
export function deriveSpawnedBoardName(
  template: RecurringBoardTemplate,
  windowStart: string,
): string {
  const trimmedName = template.name.trim();
  const windowLabel = formatTimeframeLabel(template.timeframe, windowStart);
  if (!trimmedName) return windowLabel;
  return `${trimmedName} — ${windowLabel}`;
}

/**
 * Validates a template's pool. The platform spawn driver passes the
 * resolved `poolTasks` (looked up from `seedTaskIds`); failures map to
 * user-facing "needs attention" copy on the template row.
 *
 * Pool semantics: `poolTasks.length >= fillableCellCount(...)`. The
 * spawn shuffles + slices to the cell count, so any extras become the
 * randomization pool. The earlier strict-fit `'all'` strategy was
 * dropped during the Phase 6.2 UX rework — it was just a special case
 * of the loose-fit semantics where the user happened to pick exactly
 * the cell count.
 *
 * @param template - The template being validated.
 * @param poolTasks - Resolved Tasks corresponding to `template.seedTaskIds`.
 *                    Caller is responsible for the lookup; order doesn't
 *                    matter (the spawn shuffles regardless).
 */
export function validateSpawnPool(
  template: RecurringBoardTemplate,
  poolTasks: Task[],
): SpawnPoolValidation {
  if (template.timeframe === Timeframe.CUSTOM) {
    return { ok: false, reason: 'unsupported_timeframe' };
  }
  if (template.centerSquareType === CenterSquareType.CHOSEN) {
    return { ok: false, reason: 'unsupported_center' };
  }
  if (poolTasks.some((t) => t.isDeleted)) {
    return { ok: false, reason: 'has_deleted_tasks' };
  }

  // Reject duplicate task ids in the resolved pool. The create/update
  // input schemas already reject duplicates at form-input time, but a
  // bad remote payload (e.g. an older client that didn't enforce
  // uniqueness, or a hand-crafted Firestore doc) could still slip
  // through. Without this guard `buildSpawnPlacement` would place the
  // same Task on multiple cells, contradicting the pool semantics.
  // We map this to `pool_too_small` because de-dup'd it IS too small —
  // surfacing a "needs attention" badge with that copy is closer to the
  // user-actionable story ("your pool has fewer unique tasks than this
  // board needs") than a separate `pool_has_duplicates` reason would be.
  const uniqueIds = new Set(poolTasks.map((t) => t.id));
  if (uniqueIds.size !== poolTasks.length) {
    return { ok: false, reason: 'pool_too_small' };
  }

  const required = fillableCellCount(
    template.boardSize,
    template.centerSquareType,
  );
  if (poolTasks.length < required) {
    return { ok: false, reason: 'pool_too_small' };
  }
  return { ok: true };
}

export interface BuildSpawnPlacementArgs {
  template: RecurringBoardTemplate;
  /**
   * Resolved pool. Caller must have already filtered out soft-deleted tasks
   * AND validated the pool size via `validateSpawnPool`. Behavior is
   * undefined when called with an invalid pool — defensive callers should
   * branch on `validateSpawnPool` first.
   */
  poolTasks: Task[];
  /**
   * Optional uniform `[0, 1)` RNG override. Default: `Math.random` via
   * `fisherYatesShuffle`. Tests pass a seeded RNG to make placement
   * deterministic.
   */
  rng?: () => number;
}

/**
 * Builds the cell-by-cell placement for a spawn. Returns an array of
 * length `boardSize²` where each entry is a `Task` to place on that
 * cell, or `null` for the auto-completed FREE center on
 * odd-sized boards.
 *
 * Thin wrapper over the shared `placeBoard` core (@oybc/bingo-core):
 * resolves the template's geometry / randomization / pool ordering, then
 * delegates the cell walk. The spawn path never pins a CHOSEN center
 * (`validateSpawnPool` rejects CHOSEN templates), so no `chosenCenterId`
 * is passed — a CHOSEN center that slipped through is treated as ordinary,
 * exactly as the old hand-rolled loop did.
 */
export function buildSpawnPlacement(
  args: BuildSpawnPlacementArgs,
): (Task | null)[] {
  const { template, poolTasks, rng } = args;
  return placeBoard({
    items: poolTasks,
    gridSize: template.boardSize,
    centerType: template.centerSquareType,
    randomize: template.isRandomized,
    rng,
  });
}

/**
 * Result of {@link buildRepeatBoardTemplateInput} — everything the
 * "Repeat this board…" write (docs/POOLS_RECURRING.md §Surfaces item 7)
 * needs to insert a new `RecurringBoardTemplate` directly (NOT via
 * `createRecurringBoardTemplate`, which always starts
 * `lastSpawnedWindowKey: null` — this record is born with it already set,
 * see the field doc below for why).
 */
export interface RepeatBoardTemplateInput {
  name: string;
  timeframe: Timeframe; // the CHOSEN cadence, not board.timeframe
  boardSize: BoardSize;
  centerSquareType: CenterSquareType;
  isRandomized: boolean;
  manualTaskIds: string[];
  isActive: true;
  /**
   * Pre-seeded to the CADENCE's window containing the board's start date —
   * critically NOT the board's own timeframe's window. The existing board
   * IS this window's spawn; setting this up front (rather than leaving it
   * `null`) is precisely what stops `findTemplatesPendingSpawn` from
   * generating a duplicate spawn for the same window the very next time
   * the Boards tab opens.
   */
  lastSpawnedWindowKey: string;
  /** Decode-compat snapshot — always equals `manualTaskIds` for a
   *  repeat-this-board record (there is no pool to seed from). */
  seedTaskIds: string[];
  poolIds: string[];
  removedTaskIds: string[];
}

/**
 * Builds the `RecurringBoardTemplate` insert payload for "Repeat this
 * board…" (docs/POOLS_RECURRING.md §Surfaces item 7): a one-off board
 * gaining a repeat cadence AFTER the fact, with its current placed tasks as
 * the record's entire (manual-only, zero-pool) mix.
 *
 * `lastSpawnedWindowKey` is computed from `cadence` — the newly-chosen
 * repeat cadence — evaluated against the board's OWN `startDate`, never
 * from `board.timeframe`. A one-off Tuesday DAILY board repeated Weekly
 * must key off that Tuesday's WEEK window, not the day: using the board's
 * timeframe here would produce a window key for the wrong granularity,
 * and the next `findTemplatesPendingSpawn` check would see a mismatched
 * (or already-elapsed) window and spawn a duplicate board mid-window.
 *
 * No persistence; no id/timestamps — the caller (`repeatBoardAsRecurring`)
 * fills those in as part of a single atomic write that also back-stamps
 * the source board's `spawnedFromTemplateId`.
 *
 * @param board - The source one-off board. Only `name`/`boardSize`/
 *   `centerSquareType`/`isRandomized`/`startDate` are read.
 * @param boardTaskIds - Caller-resolved: distinct, non-deleted, non-FREE-
 *   center task ids currently placed on the board (in placement order).
 * @param cadence - The newly-chosen repeat cadence (DAILY/WEEKLY/MONTHLY/
 *   YEARLY — never CUSTOM, which the picker excludes).
 * @param weekStartDay - Only relevant when `cadence === WEEKLY`.
 */
export function buildRepeatBoardTemplateInput(
  board: {
    name: string;
    boardSize: BoardSize;
    centerSquareType: CenterSquareType;
    isRandomized: boolean;
    startDate: string;
  },
  boardTaskIds: string[],
  cadence: Timeframe,
  weekStartDay: WeekStartDay,
): RepeatBoardTemplateInput {
  const { startDate: lastSpawnedWindowKey } = getTimeframeBoundaries(
    cadence,
    new Date(board.startDate),
    weekStartDay,
  );

  return {
    name: board.name,
    timeframe: cadence,
    boardSize: board.boardSize,
    centerSquareType: board.centerSquareType,
    isRandomized: board.isRandomized,
    manualTaskIds: [...boardTaskIds],
    isActive: true,
    lastSpawnedWindowKey,
    seedTaskIds: [...boardTaskIds],
    poolIds: [],
    removedTaskIds: [],
  };
}
